### PR TITLE
[AArch64][NEON] Lower fixed-width add partial reductions to dot product

### PR DESCRIPTION
**Pull request by SamTebbs33**
This PR adds lowering for fixed-width <4 x i32> and <2 x i32> partial reductions to a dot product when Neon and the dot product feature are available.

The work is by Max Beck-Jones (@DevM-uk).

### CODE REVIEW

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 1]
```diff
@@ -21781,7 +21782,8 @@ SDValue tryLowerPartialReductionToDot(SDNode *N,
              Intrinsic::experimental_vector_partial_reduce_add &&
          "Expected a partial reduction node");
 
-  if (!Subtarget->isSVEorStreamingSVEAvailable())
+  if (!Subtarget->isSVEorStreamingSVEAvailable() &&
```
Comment (david-arm) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:21785]: What if this is <4 x i32> and we don't have NEON's dot product, but we do have SVE? Will we do something sensible? If we lower fixed-width using SVE, then we should add an extra RUN line to the tests for this.

Comment (SamTebbs33) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:21785]: Thanks for spotting that missing, I've modified the check to also check for the dotprod feature when SVE is available and we're using a fixed-width vector.

File: llvm/test/CodeGen/AArch64/neon-partial-reduce-dot-product.ll [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+; CHECK-NODOTPROD-NEXT:    add v0.2s, v1.2s, v0.2s
+; CHECK-NODOTPROD-NEXT:    ret
+  %u.wide = zext <4 x i8> %u to <4 x i32>
+  %s.wide = zext <4 x i8> %s to <4 x i32>
+  %mult = mul nuw nsw <4 x i32> %s.wide, %u.wide
+  %partial.reduce = tail call <2 x i32> @llvm.experimental.vector.partial.reduce.add.v4i32.v16i32(<2 x i32> %acc, <4 x i32> %mult)
+  ret <2 x i32> %partial.reduce
+}
+
+define <4 x i32> @not_sdot(<4 x i32> %acc, <8 x i8> %u, <8 x i8> %s) #0{
```
[the remark this answers is not in the document]
Comment (SamTebbs33) [reply llvm/test/CodeGen/AArch64/neon-partial-reduce-dot-product.ll:137]: `shouldExpandPartialReductionIntrinsic` only considers the return type and `tryLowerPartialReductionToDot` checks the operand types so these do test some of those invalid operand types.

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 3]
```diff
@@ -21781,7 +21782,12 @@ SDValue tryLowerPartialReductionToDot(SDNode *N,
...
+      !Subtarget->isNeonAvailable())
+    return SDValue();
+
+  // Fixed-width requires the dotprod feature, both for Neon and SVE
+  if (!N->getValueType(0).isScalableVT() && !Subtarget->hasDotProd())
```
[the remark this answers is not in the document]
Comment (SamTebbs33) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp]: Done.

File: llvm/test/CodeGen/AArch64/neon-partial-reduce-dot-product.ll (hunk 2 above)
[the remark this answers is not in the document]
Comment (paulwalker-arm) [reply llvm/test/CodeGen/AArch64/neon-partial-reduce-dot-product.ll:137]: Restricting my comment to `shouldExpandPartialReductionIntrinsic` was an error, but what I meant is the only difference between `not_udot` vs `not_sdot` and `not_udot_narrow` vs `not_sdot_narrow` is the type of extension and so there is nothing in `not_sdot` and `not_sdot_narrow` that's not already being tested by `not_udot` and `not_udot_narrow`.

File: llvm/test/CodeGen/AArch64/partial-reduce-dot-product.ll
Comment (paulwalker-arm) [inline llvm/test/CodeGen/AArch64/partial-reduce-dot-product.ll:1]: Sorry, not sure how I missed this file before but please revert the changes because there's nothing relevant to SVE here.

If you want, perhaps just rename the original file to sve-partial-reduce-dot-product.ll.

File: llvm/test/CodeGen/AArch64/neon-partial-reduce-dot-product.ll (hunk 2 above)
[the remark this answers is not in the document]
Comment (SamTebbs33) [reply llvm/test/CodeGen/AArch64/neon-partial-reduce-dot-product.ll]: Ah I see, done.

File: llvm/test/CodeGen/AArch64/partial-reduce-dot-product.ll
Comment (SamTebbs33) [reply llvm/test/CodeGen/AArch64/partial-reduce-dot-product.ll:1]: Done, thank you.

**Review by paulwalker-arm — APPROVED** [review]
